### PR TITLE
Px4.enhancement.3760.implement auto follow me.wip [WIP do not merge]

### PR DESCRIPTION
**Pull request by catch-twenty-two**
adds ability for px4 firmware to track a target.  Implements rudimentary following of target position in navigator.

Note: this uses a new version of the mavlink library with the proposed follow target method, will not work with current mavlink library version.

https://github.com/catch-twenty-two/mavlink

Also uses new version of qground control with follow me option implemented:

https://github.com/mavlink/qgroundcontrol/pull/2817

Some concerns:

Please comment on ideas for error checking on bad position sent from tracker.  (e.g. ignore tracker if new pos vector is x meters away and instead just loiter)

Does it make sense for the "follow previous - current line" following be activated for this?  (currently I have it activated) I guess this might come in handy for intermittent GPS being sent...

In both instances ( "follow previous - current line" on and off) the result is "jumpy" if the new sp is to close to the current set point. 

Maybe this entire follow feature needs to be treated more like an instance using the position control code, except instead of manual input it gets tracker input. Or maybe it needs to be some type of hybrid of both pos control and auto line/position following code.

Anyways, just some ideas, I'd like to get other's suggestions. 

@LorenzMeier had suggested using a predictive model if we have the target velocity. (this doesn't seem to be an option on my nexus 7 as it doesn't seem to output gps velocity)

Here is a video of the current implementation, Qground control has been coded to send out sinusoidal gps coors, you can see the jumpiness I am referring to:

https://www.youtube.com/watch?v=qGC464eVI5E&feature=youtu.be

### CODE REVIEW

**Comment by devbharat** [comment]
This reminded me of the target following I had implemented in APM a few years ago (used the same fixed wing libraries (l1 with TECS) as we do now) and the issues I faced. I did it by setting the target's position as the waypoints for the 'follower' (also I think I overwrote the current target waypoint with the latest waypoint I received from the target. Otherwise, the follower's position oscillated heavily). 
This strategy 'works' but you end up with steady state error between the target(say, moving in a straight line) and the follower (who never got directly in line behind the target, was always a bit to the left or the right). This can be seen in the following pic:

dsc_0615

You can see the green plane (fixed wing UAV) is following not directly behind the leader, but to the side. This was because unlike a single static waypoint, the waypoints in this case were constantly overwritten to the new target position values, with the controller never getting the time to correct the steady state error. This made me feel that, atleast for fixedwings, tracking moving targets required something more than just setting moving waypoints from the controls point of view.

You can see a mission with 3 aircrafts daisy-chained here:
3-ac-lf-2013-09-28 13_41_29

**Comment by LorenzMeier** [comment]
Awesome! I'm not so worried about the back and forth as my intention was not to keep sending a specific setpoint all the time, but to implement a "rubber-band" controller in that navigator class which trades smoothness for accuracy and only updates the next position gradually.

What I however indeed forgot is that we need to feed a desired velocity. That should be supported by the triplet as well. Or we can go straight to the position setpoint instead. I need to think about that and maybe we can discuss this on the dev call?

**Comment by catch-twenty-two** [comment]
> What I however indeed forgot is that we need to feed a desired velocity. That should be supported by the triplet as well. Or we can go straight to the position setpoint instead. I need to think about that and maybe we can discuss this on the dev call?

Okay sounds good. Ill be on the call this morning, interested in hearing more about the controller you have in mind.

**Comment by LorenzMeier** [comment]
@catch-twenty-two Any update on the current state? Could you push your latest commits so others have a chance to review and join in?

**Comment by LorenzMeier** [comment]
You also might want to rebase. We kept improving the sim, in particular Gazebo, quite a bit.

**Comment by catch-twenty-two** [comment]
Sorry your request got lost in all the noise of my email in box.  The first iteration of the px4 side of the firmware code should be complete now.  Over all operation is a lot smoother.

File: src/modules/commander/state_machine_helper.cpp [hunk 1]
```diff
@@ -764,6 +765,27 @@ bool set_nav_state(struct vehicle_status_s *status, const bool data_link_loss_en
 		}
 		break;
 
+	case vehicle_status_s::MAIN_STATE_AUTO_FOLLOW_TARGET:
+		/* require global position and home */
+
+		if (status->engine_failure) {
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LANDENGFAIL;
+		} else if ((!status->condition_global_position_valid ||
```
Comment (catch-twenty-two) [inline src/modules/commander/state_machine_helper.cpp]: Don't require a home position

**Comment by liang-tang** [comment]
@catch-twenty-two I reviewed your code. You can use a pixhawk & mavros to send waypoint message include gps velocity. run rosrun mavros gcs_bridge, roslaunch mavros px4.launch.

**Comment by LorenzMeier** [comment]
@liang-tang You missed the intent of this PR. It does follow-me without a companion computer. So what you describe is not of interest here.

**Comment by liang-tang** [comment]
@catch-twenty-two i'm confused about the following code
`if(_pos_sp_triplet.current.velocity_valid) {
                    _vel_sp(0) += _pos_sp_triplet.current.vx;
                    _vel_sp(1) += _pos_sp_triplet.current.vy;
                }

                                if (_run_alt_control) {
                                        _vel_sp(2) = (_pos_sp(2) - _pos(2)) * _params.pos_p(2);
                                }

                                /* make sure velocity setpoint is saturated in xy*/
                                float vel_norm_xy = sqrtf(_vel_sp(0) * _vel_sp(0) +  _vel_sp(1) * _vel_sp(1));

                if (_pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_FOLLOW_TARGET) {
                    _vel_sp(0) = _pos_sp_triplet.current.vx;
                    _vel_sp(1) = _pos_sp_triplet.current.vy;
                } else if (vel_norm_xy > _params.vel_max(0)) {
                    /* note assumes vel_max(0) == vel_max(1) */
                    _vel_sp(0) = _vel_sp(0) * _params.vel_max(0) / vel_norm_xy;
                    _vel_sp(1) = _vel_sp(1) * _params.vel_max(1) / vel_norm_xy;
                }
` _pos_sp_triplet.current.velocity_valid is always valid, and _pos_sp_triplet.current.type is always SETPOINT_TYPE_FOLLOW_TARGET when at follow target mode, but _pos_sp_triplet.current.velocity_valid is really working here or covered by                     _vel_sp(0) = _pos_sp_triplet.current.vx;
                    _vel_sp(1) = _pos_sp_triplet.current.vy;?

**Comment by liang-tang** [comment]
@catch-twenty-two `                if (_pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_FOLLOW_TARGET) {
                    _vel_sp(0) = _pos_sp_triplet.current.vx;
                    _vel_sp(1) = _pos_sp_triplet.current.vy;
                } else if (vel_norm_xy > _params.vel_max(0)) {
                    /* note assumes vel_max(0) == vel_max(1) */
                    _vel_sp(0) = _vel_sp(0) * _params.vel_max(0) / vel_norm_xy;
                    _vel_sp(1) = _vel_sp(1) * _params.vel_max(1) / vel_norm_xy;
                }`

**Comment by catch-twenty-two** [comment]
@liang-tang True probably better to double check than assume if you are in follow target mode you have a velocity to use.  I have since changed this entire function to use the velocity setpoint type and made some other significant changes as well.

**Comment by catch-twenty-two** [comment]
Hello, this is ready for testing if anyone has some time available. Integration to get the system working can be a bit of a pain because it involves changes in qgc, mavlink and px4.  It will also require an Android device with OTG capability. Let me know if you're interested. Thanks!

@julianoes @tumbili: @LorenzMeier said I should request validation testing from you so we can pull this into master.

**Comment by LorenzMeier** [comment]
CI failed, fix would be great. Can you also help to get the required PR for QGC through? Then it will be available on Android automatically.

**Comment by catch-twenty-two** [comment]
Sure will do

**Comment by liang-tang** [comment]
@catch-twenty-two Testing is in process.

**Comment by catch-twenty-two** [comment]
@liang-tang Great thanks, make sure you grab the latest code, for now I have set the minimum altitude of follow me/missions to 10 meters for safety.

**Comment by catch-twenty-two** [comment]
@LorenzMeier I was able to get the CI to pass by changing the mavlink submodule to point to a temporary branch with the new follow target messages...Once the pr with the new messages are approved in the mavlink repo here:

https://github.com/mavlink/mavlink/pull/526

 can switch it back to pointing to the correct repo and it should build ok without the hack.

**Comment by LorenzMeier** [comment]
@catch-twenty-two Will start to integrate this tonight, so will be done in your morning.

**Comment by LorenzMeier** [comment]
MAVLink PR is in. Now moving on to do a final review here and get it in.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 2]
```diff
@@ -1355,18 +1358,50 @@ MulticopterPositionControl::task_main()
...
 
-				if (vel_norm_xy > _params.vel_max(0)) {
-					/* note assumes vel_max(0) == vel_max(1) */
-					_vel_sp(0) = _vel_sp(0) * _params.vel_max(0) / vel_norm_xy;
-					_vel_sp(1) = _vel_sp(1) * _params.vel_max(1) / vel_norm_xy;
+                    _vel_sp(0) = fabs(ft_vel(0)) > fabs(_vel_sp(0)) ? ft_vel(0) : _vel_sp(0);
+                    _vel_sp(1) = fabs(ft_vel(1)) > fabs(_vel_sp(1)) ? ft_vel(1) : _vel_sp(1);
+                }
+
+                // Ignore position control calculated vel set points
```
Comment (LorenzMeier) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: In the section below are changes which exceed the changes needed for follow-me. Either this is a bad merge or the conditioning is incorrect (they should be guarded by if == FOLLOW_TARGET).

**Comment by LorenzMeier** [comment]
Almost all changes look good, except for the multi rotor position control where I see changes which will affect the general flight behaviour. When this is resolved it should be good to go.

I brought in the MAVLink changes in master (c27981d), please rebase.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 2 above)
Comment (catch-twenty-two) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: The above code uses a combination of overriding the velocity set point when it is appropriate while still using the position set point when it's not so I can take advantage of the velocity braking code as the position set point error delta gets smaller.  The code below overrides the velocity set point at all times and ignores the position set point.  I was going to make a new set point type for the code below but saw that the position set point type already defined a set point type for velocity only, it just wasn't being used in the code anywhere. I used both of these set point types for a smoother transition between following using position and velocity. Let me know if you still think I need to fix something.

https://github.com/PX4/Firmware/pull/3814/files?diff=split#diff-7a824e3d1e9a5420071c8c9bbe27b8e7R366

**Comment by catch-twenty-two** [comment]
> Almost all changes look good, except for the multi rotor position control where I see changes which will affect the general flight behaviour. When this is resolved it should be good to go.
> 
> I brought in the MAVLink changes in master (c27981d), please rebase.

Thanks, I addressed your concerns in the comments.  Will do the rebase asap.

**Comment by liang-tang** [comment]
@catch-twenty-two `void FollowTarget::pause()
{
    //math::Vector<3> vel(0, 0, 0);

    _current_vel.zero();

    set_loiter_item(&_mission_item, _param_min_alt.get());

    //update_position_sp(vel);

    _previous_target_gps_pos_valid = false;
    //_current_target_motion.lat = _navigator->get_global_position()->lat;
    //_current_target_motion.lon = _navigator->get_global_position()->lon;
    //mavlink_log_info(_navigator->get_mavlink_fd(),"pause");
}
`

**Comment by catch-twenty-two** [comment]
@LorenzMeier Hey Lorenz,  I got rid of the SETPOINT_TYPE_VELOCITY statement by checking if the velocity or position setpoints are valid in the setpoint triplet.  This works the exact same way, and allows me to get rid of the statement for ignoring pos setpoint generated velocity if SETPOINT_TYPE_VELOCITY is the setpoint type.

**Comment by catch-twenty-two** [comment]
@liang-tang I took some time to make the code more robust and readable overall.  It should address all of your concerns. Were you able to fly it at all?

**Comment by liang-tang** [comment]
@catch-twenty-two I tested many times, i found it difficult to catch up with me, maybe gps data lost?

**Comment by catch-twenty-two** [comment]
Yes, that might be the case, the 3dr radio interferes with my phone gps...  I have to hold it very far away from the phone to get it to work correctly...Once it works though it works well...I'm going to play around with it some more to find the optimal placement for the radio.  I was able to ride a bike stopping and starting abruptly all over a parking lot and it stayed with in 10-15 meters of me the whole time.

**Comment by LorenzMeier** [comment]
Replaced by https://github.com/PX4/Firmware/pull/4065/files

**Pull request closed without merging**